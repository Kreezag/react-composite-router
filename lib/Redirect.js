/* eslint-env browser */

import React, { Component } from 'react';
import referringRouter, { statePropTypes } from './referring';



const ssr = typeof window !== 'object';



class Redirect extends Component {

  static propTypes = {
    state: statePropTypes
  };

  render () {
    return <noscript />;
  }

  componentWillMount () {
    if (ssr) {
      this.props.state.apply();
    }
  }

  componentDidMount () {
    if (!ssr) {
      this.props.state.apply();
    }
  }
}



export default referringRouter()(Redirect);
