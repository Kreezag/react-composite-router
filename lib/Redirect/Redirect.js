/* eslint-env browser */

import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { moveTo, calcLocation } from '../routing';
import historyPropTypes from '../historyPropTypes';



const ssr = typeof window !== 'object';



class Redirect extends Component {
  static contextTypes = {
    routerRoutesByName: PropTypes.object.isRequired,
    history: historyPropTypes()
  };

  static propTypes = {
    routeState: PropTypes.object.isRequired,
    state: PropTypes.string.isRequired,
    params: PropTypes.object,
    replace: PropTypes.bool,
    reload: PropTypes.bool
  };

  static defaultProps = {
    replace: true,
    reload: false
  };

  render () {
    return null;
  }

  componentWillMount () {
    if (ssr) {
      this._redirect();
    }
  }

  componentDidMount () {
    if (!ssr) {
      this._redirect();
    }
  }

  _redirect () {
    const { history, routerRoutesByName, routerState } = this.context;
    const { replace, reload, state, params } = this.props;

    const location = calcLocation(routerRoutesByName, state, routerState.params, params);

    moveTo(history, location, { replace, reload });
  }
}



export default connect(({ routerState }) => ({ routerState }))(Redirect);
