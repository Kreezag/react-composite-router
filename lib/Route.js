import { assertTrimmedNonEmptyString, assertRegExp } from './assert';
import { uniq, each, reduce, isString, isNumber, calcSearchString, parseSearchString } from './utils';


const calcUrlProps = (url) => ({
  path: url.replace(/[?#].*$/, ''),
  searchParams: url.replace(/^[^?]*[?]?/, '').split('&').filter(Boolean)
});


export default class Route {
  constructor (name, { url = '', params = {}, slots = {}, reroute = null } = {}) {
    assertTrimmedNonEmptyString('name', name);
    assertRegExp(/^([a-zA-Z0-9-.]+)$/, 'name', name);

    const parsedUrl = calcUrlProps(url);

    this.url = url;
    this.name = name;
    this.path = parsedUrl.path;
    this.slots = slots;
    this.params = params;
    this.reroute = reroute; // function for replace routes
    this.searchParams = parsedUrl.searchParams;
    this.pathMaskParams = [];

    this.pathMask = this.path.replace(/:([^/]+)/g, ($0, paramName) => {
      this.pathMaskParams.push(paramName);

      return `{{${paramName}}}`;
    });

    this.pathMaskExpSrc = this.path.replace(/:([^/]+)/g, () => '([^/]+)');

    this.pathMaskExp = new RegExp(`^${this.pathMaskExpSrc}$`);

    this.availableUrlParams = this.searchParams.concat(this.pathMaskParams);
    this.additionalParams = Object.keys(this.params)
      .filter((param) => !this.pathMaskParams.includes(param) && !this.searchParams.includes(param));

    this.availableParams = this.availableUrlParams.concat(this.additionalParams);




    // post checking
    if (this.availableUrlParams.length !== uniq(this.availableUrlParams).length) {
      throw new Error(`this "${this.name}" has duplication in params [${this.searchParams.join(',')}] and [${this.pathMaskParams.join(',')}]`);
    }

    const invalidParams = this.additionalParams.filter((param) => !/^[^a-zA-Z]/.test(param));

    if (invalidParams.length) {
      throw new Error(`this "${this.name}" has invalid params [${invalidParams.join(',')}]. additional params must not starts from [a-zA-Z] symbold`);
    }

    const invalidUrlParams = this.availableUrlParams.filter((param) => /^[^a-zA-Z]/.test(param));

    if (invalidUrlParams.length) {
      throw new Error(`this "${this.name}" has invalid url params [${invalidUrlParams.join(',')}]. url params must starts only from [a-z] symbols`);
    }

    each(this.params, (val, key) => {
      if (!this.availableUrlParams.includes(key)) {
        return;
      }

      if (!isString(val) && !isNumber(val)) {
        throw new Error(`this "${this.name}" has invalid param type of "${key}". must be Number or String`);
      }
    });
  }

  isMatch (location, action) {
    return this.pathMaskExp.test(location.pathname || '');
  }

  parseLocationParams (location) { // location
    const { pathname, search, hash } = location;
    let params = Object.assign({}, this.params);

    String(pathname || '').replace(this.pathMaskExp, ($0, ...args) => {
      params = this.pathMaskParams.reduce((params, paramName, index) => {
        params[paramName] = args[index];

        return params;
      }, params);

      return $0;
    });

    const searchParams = search ? parseSearchString(String(search || '').replace(/^\?/, '')) : {};

    params = reduce(searchParams, (params, paramValue, paramName) => {
      if (this.searchParams.includes(paramName)) {
        params[paramName] = paramValue;
      }

      return params;
    }, params);

    return params;
  }

  calcLocation (params) {
    const pathname = this.pathMask.replace(/{{([^}]+)}}/g, ($0, paramName) => {
      if (params.hasOwnProperty(paramName)) {
        return params[paramName];
      }

      throw new Error(`Route param "${paramName}" was not defined for route "${this.name}"`);
    });

    const searchParams = this.searchParams.reduce((search, searchParam) => {
      if (params.hasOwnProperty(searchParam) && params[searchParam] !== this.params[searchParam]) {
        search[searchParam] = params[searchParam];
      }

      return search;
    }, {});

    const location = { pathname, hash: '', search: '' };

    if (searchParams) {
      location.search = calcSearchString(searchParams);
    }

    return location;
  }

  calcParams (stateParams, currentStateParams = {}, resetParams = false) {
    return this.availableParams.reduce((params, paramName) => {
      if (stateParams.hasOwnProperty(paramName)) {
        params[paramName] = stateParams[paramName];
      } else if (!resetParams && currentStateParams.hasOwnProperty(paramName)) {
        params[paramName] = currentStateParams[paramName];
      }

      return params;
    }, Object.assign({}, this.params));
  }

  static createRootRoute (name, definition) {
    const sections = String(name).split('.');

    if (sections.length !== 1) {
      throw new Error(`state "${name}" has invalid format`);
    }

    return new Route(name, definition);
  }

  createChildRoute (name, { url = '', params = {}, slots = {}, reroute = null } = {}) {
    const sections = String(name).split('.');

    if (`${this.name}.${sections[sections.length - 1]}` !== name) {
      throw new Error(`child state name (${name}) must starts from parent name "${this.name}"`);
    }

    const { searchParams, path } = calcUrlProps(url);

    slots = Object.assign({}, this.slots, slots);
    params = Object.assign({}, this.params, params);

    const urlPath = `${this.path.replace(/\/$/, '')}/${path.replace(/^\//, '')}`;
    const urlSearch = `${[ ...this.searchParams, ...searchParams ].join('&')}`;

    url = `${urlPath}${urlSearch ? `?${urlSearch}` : ''}`;

    return new Route(name, { slots, params, url, reroute });
  }
}
