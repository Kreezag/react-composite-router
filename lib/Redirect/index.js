/* eslint-env browser */

import React from 'react';
import Redirect from './Redirect';



export const redirectRender = (state, params = {}) => () => (<Redirect state={state} params={params} />); // eslint-disable-line



export default Redirect;
