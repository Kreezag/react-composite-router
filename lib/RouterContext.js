import once from 'lodash/once';
import forEach from 'lodash/forEach';
import isEqual from 'lodash/isEqual';
import { assertAvailableProps } from './assert';
import FakeHistory from './FakeHistory';
import EE from './EE';
import Route from './Route';



export default class RouterContext {
  constructor () {
    this.history = new FakeHistory();

    this._listenHistory(this.history);

    this._stateEmitter = new EE();

    this.state = {
      name: '',
      params: {},
      composition: {}
    };

    this.location = { // there is separated location object because history might be in inconsistent state with route
      pathname: '',
      hash: '',
      search: ''
    };

    this.routes = [];
    this.routesByName = {};

    this._hooks = [];
  }

  destroy () {
    this._historyCleanupListener();
    this._stateEmitter.destroy();
    this.location = {};
    this.history = null;
    this.state = {};
    this.routes = [];
    this.routesByName = {};
    this._hooks = [];
  }

  searchStateByLocation (location, action) {
    const route = this.routes.find((route) => route.isMatch(location, action));

    if (!route) {
      return null;
    }

    return {
      name: route.name,
      params: route.parseLocationParams(location)
    };
  }

  setHistory (history) {
    this.history = history;

    this.setLocation(history.location, history.action);

    this._listenHistory(history);
  }

  _listenHistory (history) {
    if (this._historyCleanupListener) {
      this._historyCleanupListener();
    }

    this._historyCleanupListener = once(history.listen((location, action) => {
      this.setLocation(location, action);
    }));
  }

  getLocation () {
    return this.location;
  }

  appendHook (hookFunction) {
    this._hooks = [ ...this._hooks, hookFunction ]; // subscribe

    return once(() => {
      this._hooks = this._hooks.filter((callback) => callback !== hookFunction); // unsubscribe
    });
  }

  applyHooks (nextStateName, nextStateParams) {
    return Promise.all(this._hooks.map((hook) => hook(nextStateName, nextStateParams)));
  }

  setLocation (location, action) {
    const state = this.searchStateByLocation(location, action);

    if (!state) {
      window.console && console.error && console.error(`Router: state is not defined (location: ${JSON.stringify(location)})`); // eslint-disable-line no-console, no-unused-expressions
      return; // not found. rollback to previous state
    }

    if (!this._hooks.length) {
      this.setState(state.name, state.params);
    } else {
      this.applyHooks(state.name, state.params).then(() => {
        this.setState(state.name, state.params);
      });
    }
  }

  setRoutes (routes) {
    forEach(routes, (route) => {
      if (!(route instanceof Route)) {
        throw new TypeError('object is not instance of Route');
      }
    });

    this.routesByName = routes;

    this.routes = Object.keys(routes)
      .sort((name1, name2) => {
        const segmentsLen1 = name1.split('.').length;
        const segmentsLen2 = name2.split('.').length;

        if (segmentsLen1 > segmentsLen2) {
          return -1;
        }

        return segmentsLen1 < segmentsLen2 ? 1 : 0;
      })
      .map((name) => routes[name]);
  }

  setState (name, params = {}) {
    if (this.state.name === name && isEqual(this.state.params, params)) {
      window.console && console.info && console.info('Router: state is the same'); // eslint-disable-line no-console, no-unused-expressions
      return; // state is equal. nothing to do
    }

    if (!this.routesByName.hasOwnProperty(name)) {
      throw new Error(`invalid state name "${name}"`);
    }

    this.state = {
      name,
      params,
      composition: this.routesByName[name].slots
    };

    this.location = this.calcLocation(name, params);

    this._stateEmitter.publish(name, params);
  }

  onStateChange (listenerFunc) {
    return this._stateEmitter.subscribe(listenerFunc);
  }

  moveTo (location, { reload = false, replace = false } = {} /*{ replace, reload }*/) {
    if (!location) {
      throw new Error('location is empty');
    }

    if (!reload) {
      if (replace) {
        this.history.replace(location);
      } else {
        this.history.push(location);
      }

      return;
    }

    window.location.href = this.history.createHref(location);
  }

  createHref (location) {
    return this.history.createHref(location);
  }

  calcLocation (stateName, stateParams, { reset = false } = {}) {
    if (!this.routesByName.hasOwnProperty(stateName)) {
      throw new Error(`RouterContext: state "${stateName}" not found`);
    }

    const route = this.routesByName[stateName];
    const params = route.calcParams(stateParams, this.state.params, reset);

    if (process.env.NODE_ENV !== 'production') { // eslint-disable-line no-process-env
      assertAvailableProps(route.availableParams, 'route params', params);
    }

    if (route.reroute) {
      const newState = route.reroute({ name: route.name, params }, { name: this.state.name, params: this.state.params });

      if (newState !== undefined) {
        if (this.routesByName.hasOwnProperty(newState.name)) {
          const newRoute = this.routesByName[newState.name];

          return newRoute.calcLocation(Object.assign({}, newRoute.params, newState.params || {}));
        }

        throw new Error('invalid route after rerouting');
      }
    }

    return route.calcLocation(params);
  }

  getState () {
    return this.state;
  }
}
