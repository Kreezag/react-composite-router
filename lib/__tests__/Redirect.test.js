/* eslint-env browser, jest */
/* eslint-disable react/display-name */

import React from 'react';
import renderer from 'react-test-renderer';
import WithContext from 'react-with-context';
import Redirect from '../Redirect';
import RouterContext from '../RouterContext';
import Route from '../Route';
import history from './__mocks__/history';

describe('rendering', () => {
  it('should render something', () => {
    const routerContext = new RouterContext();

    routerContext.setRoutes({
      state: new Route('state', { url: '/', slots: { some: () => <span>Hello</span> } })
    });

    routerContext.setHistory(history({ pathname: '/' }));

    const component = renderer.create(
      <WithContext context={{ router: routerContext }}>
        <Redirect state="state" />
      </WithContext>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
