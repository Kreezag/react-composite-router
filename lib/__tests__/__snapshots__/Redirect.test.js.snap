exports[`rendering should render something 1`] = `null`;
